### ISSUE TITLE
Using the output of a shared library as an input into a generator or custom target

### ISSUE THREAD

**Issue by tbodt** [mesonbuild/meson#1848]
My build needs to build a shared library from some sources, and then feed that shared library into a program that generates a C file. I've tried something like `gen.process(library.full_path())` but I get an error from ninja, something like `error: /home/tbodt/blah/blah/build/library.so, needed by library.so.c, missing and no known rule to make it`.

**Comment by jpakkane** [comment]
If you are processing only one file, use a custom_target instead. Generator is for batch converting many files with the same conversion rule.

**Comment by tbodt** [comment]
The real problem here is I'm trying to use a shared library outputted by one build phase as input into another. Is there any way to do that?

**Comment by nirbheek** [comment]
You can do that with `custom_target()`.

**Comment by marc-h38** [comment]
@tbodt can this be closed?

**Comment by tbodt** [comment]
This is still a thing that can't be done with meson. I ended up using a custom target that just invokes the C compiler, but then you don't get to use meson's knowledge of how to invoke the C compiler.

**Comment by marc-h38** [comment]
>  I ended up using a custom target that just invokes the C compiler, but then you don't get to use meson's knowledge of how to invoke the C compiler.

Would a generator know how to invoke the C compiler?

> My build needs to build a shared library from some sources, and then feed that shared library into a program that generates a C file.

Have you considered chaining multiple meson.build and/or (sub)projects?

**Comment by tbodt** [comment]
> Would a generator know how to invoke the C compiler?

No, I needed to define the compiler flags myself. This turned out to be better for my use case because the library needs to be compiled in a very specific way.

> Have you considered chaining multiple meson.build and/or (sub)projects?

You mean a generator that recursively invokes meson? That doesn't sound very sane.

**Comment by marc-h38** [comment]
> > Would a generator know how to invoke the C compiler?

> No,

So what would make a generator better than a custom_target then?

> You mean a generator that recursively invokes meson? 

No, I mean having as the top level target of a sub-meson.build  either the shared library or the C code generator it is "fed" (linked?) into. That program seems complex enough to deserve its own build.

**Comment by tbodt** [comment]
> So what would make a generator better than a custom_target then?

Oh, I thought you asked whether a custom_target would know how to invoke the C compiler. Neither would. In my mind the only difference between a generator and custom target is a generator can be used more than once.

> No, I mean having as the top level target of a sub-meson.build either the shared library or the C code generator it is "fed" (linked?) into.

Can you give more details on what that would look like? I don't see a way to do this without invoking a second copy of meson.

For context, this program is a VDSO for my Linux compatibility layer for iOS. https://github.com/tbodt/ish

**Comment by marc-h38** [comment]
> Oh, I thought you asked whether a custom_target would know how to invoke the C compiler

I did, but only for comparison with a generator.

> Neither would.

This github issue requests that generator do something, but the answer so far has been "use a custom_target" instead, and you did, and it worked! That's why I asked whether the issue can be closed.

> In my mind the only difference between a generator and custom target is a generator can be used more than once.

Do you need this more than once or do you think anyone would? If yes, doesn't a foreach+custom_target() solve this? https://groups.google.com/d/msg/mesonbuild/m7aNH24zkw0/6Clkkkw7BQAJ

**Comment by marc-h38** [comment]
> Can you give more details on what that would look like?

Unfortunately not as I never did this myself but I would start from these:
https://mesonbuild.com/FAQ.html#should-i-use-subdir-or-subproject
https://mesonbuild.com/Wrap-dependency-system-manual.html

>  I don't see a way to do this without invoking a second copy of meson.

You mean a second `meson.build` file? (which is what I meant)

**Comment by tbodt** [comment]
Hmm I think there's a bit of confusion about what the custom target is used for. There's two steps: compiling the VDSO library and converting it into a C file. The original request was to allow the output of a shared_library (first step) to be the input of a generator (second step). My use case is solved by using a custom_target for the first step, not the second step. Using the output of a shared_library or other executable target as the input of a generator is, I believe, still impossible.

> You mean a second meson.build file? (which is what I meant)

I don't think that would actually work here. What I wanted to do was this:

```
foo = shared_library('foo', ['foo.c'])
bar_generator = generator(...)
bar = bar_generator.process(foo)
```

but that didn't work because executable targets weren't allowed as input to generator.process, and similarly with custom_target input. (I haven't tried in a while so maybe this works today, in which case this issue should be closed.) My understanding is that subprojects and subdirectories are just a way to structure your code and won't make a difference in runtime errors like those.

**Comment by xclaesse** [comment]
I think this type of use-case can be done with that PR: https://github.com/mesonbuild/meson/pull/4436. If meson generates an -uninstalled.pc file for your library, then your script can use pkg-config to get all the info needed to use that library, like include paths within source tree, other libraries to link, etc.

**Comment by marc-h38** [comment]
> My use case is solved by using a custom_target for the first step, not the second step.

I read it at least five times and I'm fairly confident the very first comment above (and documentation) suggested to use a `custom_target` for the *second* step... so that's what I meant the whole time.

> that didn't work because executable targets weren't allowed as input to generator.process, *and similarly with custom_target input*

Thanks, that answers my question almost 10 comments ago :-)

> (I haven't tried in a while so maybe this works today,

**Comment by thisismiller** [comment]
This prevents using any sort of custom harness for executing tests.  I'll use `ionice` as the example (but in my case, it's run the binary under system qemu which takes a few additional steps).  The standard build system workaround when the build system wants a single target to run as the test is to write a new shell script that looks like:
```
#!/bin/bash
ionice ./path/to/executable
```
The easy way to get this hooked into meson would then be defining a generator which transforms executables into executables-which-are-actually-shell-wrappers-that-depend-on-the-input:
```
write_test_script = find_program('write_test_script')
gen_ionice_test = generator(
  write_test_script,
  arguments : ['@INPUT@', '@OUTPUT@'],
  output : ['@BASENAME@_test.sh'],
)
```
And then one would do `test('niced test', gen_ionice_test.process(test_exe), depends : [test_exe])`

This... resoundingly doesn't work, because test() can't take the output of a generator even if it did, but getting generators able to take executables and produce something else would be the first step.  (And user definable functions / macros would be great to then remove the boilerplate of calling a few generators and setting up a test() target with the right deps , but apparently that's a hard no in meson.)
